### PR TITLE
Initial usage profile integration

### PR DESCRIPTION
**Pull request by bastjan**
## Summary

Contains the initial implementation for the usage profiles:

* `ZoneUsageProfileSyncReconciler`
  * Syncs the UsageProfile from the `control-api` to the local cluster.
  * Uses normal Reconcile loop, setup with `SetupWithManagerAndForeignCluster`
  * Has a local cache of profiles (ZoneUsageProfile)
* `ZoneUsageProfileApplyReconciler`
  * Applies selected (or all) UsageProfiles to all organization namespaces
    * Dynamic selection is in further tickets
  * Watches UsageProfiles, namespaces with organisation label, and dynamically watches objects it applies
  * Allows per-namespace overrides
    * see `transformers.NewResourceQuotaTransformer`
  * Tested using envtest, including dynamic watches
* `NamespaceQuotaValidator` webhook
  * Counts namespaces when creating new namespaces or projects and denies if limit is reached
  * Uses the users default org if not found in created

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Update tests.

### CODE REVIEW

File: scratchpad-help-sa-token.md
Comment (bastjan) [inline scratchpad-help-sa-token.md:1]: Will update if installation on RMA0 is finished.

**Review by simu — COMMENTED** [review]
LGTM overall, some comments & questions inline.

High-level, I'm missing using the UsageProfile's `spec.namespaceCount` in the new validating webhook (also noted inline).

File: webhooks/namespace_quota_validator.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+	// count namespaces for organization
+	var nsList corev1.NamespaceList
+	if err := v.Client.List(ctx, &nsList, client.MatchingLabels{
+		v.OrganizationLabel: organizationName,
+	}); err != nil {
+		l.Error(err, "error while listing namespaces")
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	if len(nsList.Items) >= v.DefaultNamespaceCountLimit {
```
[the remark this answers is not in the document]
Comment (bastjan) [reply webhooks/namespace_quota_validator.go]: Yep this went under while implementing. Was already on it while you reviewed 😅 Implemented now.

File: webhooks/namespace_quota_validator_test.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+						},
+					},
+				},
+			},
+			object: &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test",
+				},
+			},
+			allowed: false,
```
[the remark this answers is not in the document]
Comment (bastjan) [reply webhooks/namespace_quota_validator_test.go]: Now matching organization in the message.

File: webhooks/namespace_quota_validator.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+		l.Error(err, "failed to decode request")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	organizationName, _, err := unstructured.NestedString(rawObject.Object, "metadata", "labels", v.OrganizationLabel)
+	if err != nil {
+		l.Error(err, "error while fetching organization label")
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	// get the organization name from the user if it is not set on the namespace/projectrequest
```
[the remark this answers is not in the document]
Comment (bastjan) [reply webhooks/namespace_quota_validator.go:77]: I'm actually not sure; just saw it in the kyverno policies. I found one non-official reference where mutating webhooks come first, which would make sense. https://techblog.cisco.com/blog/dark-side-of-kubernetes-admission-webhooks

I log if the org came from the user. If we see it's not necessary we can remove it.

[the remark this answers is not in the document]
Comment (simu) [reply webhooks/namespace_quota_validator.go:79]: IIRC in Kyverno it was necessary, because there's no well-defined order in which policies are processed.

**Review by simu — APPROVED** [review]
LGTM now

**Pull request merged**